K2Field.K2NE.ServiceBroker/ServiceObjects/ExcelServices/ExcelImportServiceSO.cs

```diff
@@ -0,0 +1,528 @@
+                using (smoClientServer.Connection)
+                {
+                    SOC.SmartObject soImport = smoClientServer.GetSmartObject(smartObject);
+
+                    // populate an array of column names
+                    soColumnNames = new string[soImport.Properties.Count];
+                    for (int i = 0; i < soImport.Properties.Count; i++)
+                    {
+                        soColumnNames[i] = soImport.Properties[i].Name.ToLower();
+                    }
+
+                    var arrMatchingCols = dsColumnNames.Join(soColumnNames, dscol => dscol, socol => socol, (dscol, socol) => socol).ToList();
+
+                    if (arrMatchingCols.Count == 0)
+                    {
+                        result = "No matching columns found in SmartObject and Excel Data.";
+                        return result;
+                    }
+
+                    // Bulk Insert into SmartObject
+                    try
+                    {
+                        using (SOC.SmartObjectList inputList = new SOC.SmartObjectList())
+                        {
+                            // If the CreateMethodName is not specified, use the default value of "Create"
+                            if (createMethodName == null || createMethodName == string.Empty)
+                            {
+                                soImport.MethodToExecute = "Create";
+                            }
+                            else
+                            {
+                                soImport.MethodToExecute = createMethodName;
+                            }
+
+                            string sTransactionIDName = string.Empty;
+                            if (!(transactionIDName == null || transactionIDName == string.Empty))
+                            {
+                                sTransactionIDName = transactionIDName.ToLower();
+                            }
+
+                            string sTransactionIDValue = string.Empty;
+                            if (!(transactionIDValue == null || transactionIDValue == string.Empty))
+                            {
+                                sTransactionIDValue = transactionIDValue;
+                            }
+
+                            if (arrMatchingCols.Contains(sTransactionIDName))
+                            {
+                                throw new ApplicationException(Resources.ExcelImportTransactionIDNameExist); 
+                            }
+
+                            if (arrMatchingCols.Contains("FQN"))
+                            {
+                                throw new ApplicationException(Resources.ExcelImportFQNExist);
+                            }
+
+                            foreach (DataRow dr in dt.Rows)
+                            {
+                                SOC.SmartObject newSmartObject = soImport.Clone();
```

Comment:
Why do we do a clone() here? I worry that it leaks memory...

Reply:
Discussed.